network/tcp.go

```diff
@@ -0,0 +1,422 @@
+	if c.closed == true {
+		return nil
+	}
+	err := c.conn.Close()
+	c.closed = true
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// handleError produces the higher layer error depending on the type
+// so user of the package can know what is the cause of the problem
+func handleError(err error) error {
+
+	if strings.Contains(err.Error(), "use of closed") || strings.Contains(err.Error(), "broken pipe") {
+		return ErrClosed
+	} else if strings.Contains(err.Error(), "canceled") {
+		return ErrCanceled
+	} else if err == io.EOF || strings.Contains(err.Error(), "EOF") {
+		return ErrEOF
+	}
+
+	netErr, ok := err.(net.Error)
+	if !ok {
+		return ErrUnknown
+	}
+	if netErr.Temporary() {
+		return ErrTemp
+	} else if netErr.Timeout() {
+		return ErrTimeout
+	}
+	return ErrUnknown
+}
+
+// TCPListener is the underlying implementation of
+// Host using Tcp as a communication channel
+type TCPListener struct {
+	// the underlying golang/net listener
+	listener net.Listener
+	// the close channel used to indicate to the listener we want to quit
+	quit chan bool
+	// quitListener is a channel to indicate to the closing function that the
+	// listener has actually really quit
+	quitListener  chan bool
+	listeningLock sync.Mutex
+	listening     bool
+
+	// closed is used to tell the listen routine to return immediatly if a
+	// Stop() have been made (concurrency into place here)
+	closed bool
+
+	// listening addr (actual). Useful for listening on :0 port
+	addr net.Addr
+}
+
+// NewTCPListener returns a Listener. This function tries to bind to the given
+// address already.It returns the listener and an error if one occured during
+// the binding. A subsequent call to Address() will give the real listening
+// address (useful if you set it to port :0).
+func NewTCPListener(addr Address) (*TCPListener, error) {
+	l := &TCPListener{
+		quit:         make(chan bool),
+		quitListener: make(chan bool),
+	}
+	return l, l.bind(addr)
+}
+
+func (t *TCPListener) bind(addr Address) error {
+	t.listeningLock.Lock()
+	defer t.listeningLock.Unlock()
+	if t.listening == true {
+		return errors.New("Already listening")
+	}
+	global, _ := GlobalBind(addr.NetworkAddress())
+	for i := 0; i < MaxRetryConnect; i++ {
+		ln, err := net.Listen("tcp", global)
+		if err == nil {
+			t.listener = ln
+			break
+		} else if i == MaxRetryConnect-1 {
+			return errors.New("Error opening listener: " + err.Error())
+		}
+		time.Sleep(WaitRetry)
+	}
+	t.addr = t.listener.Addr()
+	return nil
+}
+
+// Listen implements the Listener interface
+func (t *TCPListener) Listen(fn func(Conn)) error {
+	receiver := func(tc *TCPConn) {
+		go fn(tc)
+	}
+	return t.listen(receiver)
+}
+
+// listen is the private function that takes a function that takes a TCPConn.
+// That way we can control what to do of the TCPConn before returning it to the
+// function given by the user. fn is called in the same routine.
+func (t *TCPListener) listen(fn func(*TCPConn)) error {
```

Comment:
Again, this is in prevision of future Conn that relies on TCP. Within PURBConn, I want to be able to access the `TCPConn` , and not an opaque `Conn` interface.